### PR TITLE
[ruby19] ruby 1.9 compatibility

### PR DESCRIPTION
**Pull request by ufoot**
The core `ddtrace` gem is 1.9 compatible. 1.8 compatibility would really break current code and force us to rewrite things with an old, akward syntax. It is tested against 1.9.3 under the CI, but the gem only requires 1.9.1 as there's no fundamental reason for it not to run with 1.9.1 (beside 1.9.1 not being able to install https gems under the CI env).

Most integrations work, except: 

  - Rails 3 when used with PG (requires 2.0)
  - Rails 4 (requires 2.1)
  - Rails 5 (requires 2.2)
  - Sidekiq (requires 2.0)

The following integrations required minor patches:

- elasticsearch
- http
- sinatra

### CODE REVIEW

File: Appraisals [hunk 1]
```diff
@@ -1,90 +1,102 @@
-if RUBY_VERSION < "2.4.0"
-  appraise "rails3-postgres" do
-    gem "test-unit"
-    gem "rails", "3.2.22.5"
-    gem "pg", platform: :ruby
-    gem 'activerecord-jdbcpostgresql-adapter', platform: :jruby
+if RUBY_VERSION < '2.4.0'
+  if RUBY_VERSION >= '1.9.1'
```
Comment (GeorgeErickson) [inline Appraisals]: There are versions of sidekiq for ruby 1.9

File: circle.yml [hunk 2]
```diff
@@ -5,8 +5,11 @@ machine:
     # FIXME: Disabled $JRUBY_VERSIONS tests because of a Java incompatibility
```
Comment (GeorgeErickson) [inline circle.yml]: not really part of the pr, but I really think jruby should be supported (and should be much easier with 1.9 support).

File: lib/ddtrace/contrib/elasticsearch/patcher.rb [hunk 3]
```diff
@@ -25,6 +40,63 @@ def patch
           @patched
         end
 
+        # rubocop:disable Metrics/MethodLength
+        def patch_elasticsearch_transport_client
```
Comment (GeorgeErickson) [inline lib/ddtrace/contrib/elasticsearch/patcher.rb]: It might be good to make sure this works with a few of the [major transport implementations](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-transport#transport-implementations) .  e.g. (patron and faraday)

File: lib/ddtrace/contrib/http/patcher.rb [hunk 4]
```diff
@@ -18,7 +18,7 @@ def should_skip_tracing?(req, address, port, transport, pin)
         # we don't want to trace our own call to the API (they use net/http)
         # when we know the host & port (from the URI) we use it, else (most-likely
         # called with a block) rely on the URL at the end.
-        if req.uri
+        if req.respond_to?(:uri) && req.uri
```
Comment (GeorgeErickson) [inline lib/ddtrace/contrib/http/patcher.rb]: What happens if I don't want to use net/http?

File: Appraisals (hunk 1 above)
Comment (ufoot) [inline Appraisals]: Yes, AFAIR this is "just a hole" in the test matrix, it probably requires some versioning in Appraisal (like we do with `Rails(3|4|5)`) but will dig into it.

File: lib/ddtrace/contrib/http/patcher.rb (hunk 4 above)
Comment (ufoot) [inline lib/ddtrace/contrib/http/patcher.rb]: Mmm, it's really a requirement at this point. We have plans to offer UDP connectivity for the agent (the main problem is the payload size, it's not uncommon to have traces which are 100k big, and need to be transferred at once) but for now installing our client lib requires `net/http` and uses it. Do you feel this is a problem?

Comment (GeorgeErickson) [inline lib/ddtrace/contrib/http/patcher.rb]: Some people like to audit/log all requests, so they converge on something like faraday/patreon to provide these hooks.

It's a fairly minor issue until someone asks for it.

Comment (ufoot) [inline lib/ddtrace/contrib/http/patcher.rb]: Gotcha, thanks for spotting this.
